opencensus/exporters/trace/zipkin/internal/zipkin_exporter.cc

```diff
@@ -0,0 +1,365 @@
+    writer->Key("tags");
+    writer->StartObject();
+    for (const auto &attribute : span.attributes()) {
+      writer->String(attribute.first);
+      writer->String(AttributeValueToString(attribute.second));
+    }
+    writer->EndObject();
+  }
+
+  writer->Key("timestamp");
+  writer->Int64(absl::ToUnixMicros(span.start_time()));
+
+  writer->Key("duration");
+  writer->Int64(absl::ToInt64Microseconds(span.end_time() - span.start_time()));
+
+  writer->EndObject();
+}
+
+// JSON encoding
+std::string EncodeJson(
+    const std::vector<::opencensus::trace::exporter::SpanData> &spans,
+    const ZipkinExporterOptions::Service &service) {
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+
+  writer.StartArray();
+  for (const auto &span : spans) {
+    SerializeJson(span, service, &writer);
+  }
+  writer.EndArray();
+  return buffer.GetString();
+}
+
+std::string GetIpAddress(ZipkinExporterOptions::AddressFamily af_type) {
+  struct ifaddrs *if_address_list;
+  struct ifaddrs *if_address;
+
+  getifaddrs(&if_address_list);
+  for (if_address = if_address_list; if_address != nullptr;
+       if_address = if_address->ifa_next) {
+    if (if_address->ifa_addr == nullptr) {
+      continue;
+    }
+    if (if_address->ifa_addr->sa_family == AF_INET) {
+      if (af_type != ZipkinExporterOptions::AddressFamily::kIpv4) continue;
+      char address[INET_ADDRSTRLEN];
+      inet_ntop(AF_INET,
+                &(reinterpret_cast<struct sockaddr_in *>(if_address->ifa_addr)
+                      ->sin_addr),
+                address, INET_ADDRSTRLEN);
+      std::string ipv4_address(address);
+      if (ipv4_address.compare(ipv4_loopback) == 0)
+        continue;
+      else
+        return ipv4_address;
+    } else if (if_address->ifa_addr->sa_family == AF_INET6) {
+      if (af_type != ZipkinExporterOptions::AddressFamily::kIpv6) continue;
+      char address[INET6_ADDRSTRLEN];
+      inet_ntop(AF_INET6,
+                &(reinterpret_cast<struct sockaddr_in6 *>(if_address->ifa_addr)
+                      ->sin6_addr),
+                address, INET6_ADDRSTRLEN);
+      std::string ipv6_address(address);
+      if (ipv6_address.compare(ipv6_loopback) == 0)
+        continue;
+      else
+        return ipv6_address;
+    }
+  }
+
+  if (af_type == ZipkinExporterOptions::AddressFamily::kIpv4)
+    return ipv4_loopback;
+  else
+    return ipv6_loopback;
+}
+
+class CurlEnv {
+ public:
+  CurlEnv() { curl_global_init(CURL_GLOBAL_DEFAULT); }
+  ~CurlEnv() { curl_global_cleanup(); }
+};
+
+void CurlSendMessage(const uint8_t *data, const ZipkinExporterOptions &options,
+                     size_t size, CURL *curl, struct curl_slist *headers,
+                     char *err_msg) {
+  CURLcode res;
+
+  if (CURLE_OK !=
+      (res = curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, err_msg))) {
+    // Failed to set curl error buffer.
+    std::cerr << curl_easy_strerror(res);
```

Comment:
Just "return res" here and the rest of the function and have the caller handle logging.

Reply:
Done.